main.py

```diff
@@ -37,129 +43,122 @@
+    global all_texts
+    docs_references = search(inp, "LangChain_agent_docs", callbacks=callbacks)
+    # repo_references = search(inp, "LangChain_agent_repo", callbacks=callbacks)
+    all_references = docs_references
+    all_references_sources = [r for r in all_references if r.metadata['source']]
+
+    sources = search(inp, "LangChain_agent_sources", callbacks=callbacks)
+    sources_docs = [doc for doc in all_texts if doc.metadata['source'] in [source.page_content for source in sources]]
+    combined_sources = sources_docs + all_references_sources
+    
+    return [doc.page_content for doc in combined_sources]
+
+def get_tools():
+    langchain_tool = Tool(
+        name="Documentation",
+        func=search_everything,
+        description="useful for when you need to refer to LangChain's documentation, for both API reference and codebase",
+    )
+    ALL_TOOLS = [langchain_tool]
+    
+    return ALL_TOOLS
+
+def get_agent(llm, chat_history: Optional[list] = None):
+
+    system_message = SystemMessage(
+            content=(
+                "You are an expert developer who is tasked with scouring documentation to answer question about LangChain. "
+                "Answer the following question as best you can. "
+                "Be inclined to include CORRECT Python code snippets if relevant to the question. If you can't find the answer, DO NOT hallucinate. Just say you don't know. "
+                "You have access to a LangChain knowledge bank retriever tool for your answer but know NOTHING about LangChain otherwise. "
+                "Always provide articulate detail to your action input. "
+                "You should always first check your search tool for information on the concepts in the question. "
+                "For example, given the following input question:\n"
+                "-----START OF EXAMPLE INPUT QUESTION-----\n"
+                "What is the transform() method for runnables? \n"
+                "-----END OF EXAMPLE INPUT QUESTION-----\n"
+                "Your research flow should be:\n"
+                "1. Query your search tool for information on 'Transform() method' to get as much context as you can about it. \n"
+                "2. Then, query your search tool for information on 'Runnables' to get as much context as you can about it. \n"
+                "3. Answer the question with the context you have gathered."
+                "For another example, given the following input question:\n"
+                "-----START OF EXAMPLE INPUT QUESTION-----\n"
+                "How can I use vLLM to run my own locally hosted model? \n"
+                "-----END OF EXAMPLE INPUT QUESTION-----\n"
+                "Your research flow should be:\n"
+                "1. Query your search tool for information on 'vLLM' to get as much context as you can about it. \n"
+                "2. Answer the question as you now have enough context."
+    ))
+    
     if chat_history:
```

Comment:
Same question as above - I think you can always just pass a memory in and if there's no chat messages it'll just populate `MessagesPlaceholder` with nothing

Reply:
fixed